src/service/future_task.rs

```diff
@@ -97,48 +87,41 @@ impl Stream for FutureTaskManager {
     type Item = ();
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        let mut task_finished = false;
-        let mut id_finished = false;
-        for _ in 0..128 {
-            if self.shutdown.load(Ordering::SeqCst) {
-                debug!("future task finished because service shutdown");
-                return Poll::Ready(None);
-            }
+        if self.shutdown.load(Ordering::SeqCst) {
+            debug!("future task finished because service shutdown");
+            return Poll::Ready(None);
+        }
 
+        loop {
             match Pin::new(&mut self.task_receiver).as_mut().poll_next(cx) {
                 Poll::Ready(Some(task)) => self.add_task(task),
                 Poll::Ready(None) => {
                     debug!("future task receiver finished");
                     return Poll::Ready(None);
                 }
                 Poll::Pending => {
-                    task_finished = true;
                     break;
                 }
             }
         }
 
-        for _ in 0..64 {
-            if self.shutdown.load(Ordering::SeqCst) {
-                debug!("future task finished because service shutdown");
-                return Poll::Ready(None);
-            }
-
+        loop {
             match Pin::new(&mut self.id_receiver).as_mut().poll_next(cx) {
                 Poll::Ready(Some(id)) => self.remove_task(id),
                 Poll::Ready(None) => {
                     debug!("future task id receiver finished");
                     return Poll::Ready(None);
                 }
                 Poll::Pending => {
-                    id_finished = true;
                     break;
                 }
             }
         }
 
-        if !task_finished || !id_finished {
-            self.set_delay(cx);
+        // double check here
+        if self.shutdown.load(Ordering::SeqCst) {
+            debug!("future task finished because service shutdown");
+            return Poll::Ready(None);
```

Comment:
If you set a timer, when the timer reach the end, there maybe generate more timers.

So, we need to check `shutdown` variable every time when invoke a `poll_next ()`. Otherwise it can be a dead loop.

Reply:
When the manager is closed, all futures are closed and do not need to be checked every time, sync this stuff is a relatively time-consuming thing